### PR TITLE
[RFC] Add a libuv wrapper for fsync

### PR DESCRIPTION
**Pull request by sethjackson**
We have a CMake define `HAVE_FSYNC` to check if we have `fsync()` support or not.
Since libuv provides `uv_fs_fsync` we can use libuv instead.
So add an `os_fsync` wrapper and remove `HAVE_FSYNC`.

cc @equalsraf @justinmk @ZyX-I

### CODE REVIEW

**Comment by justinmk** [comment]
@sethjackson This and other recent PRs failed because of luajit.org connection failures which only affected our builds (seems like they are blocking travis). Rebasing on master will fix this.

**Comment by sethjackson** [comment]
Yup. Thanks for fixing that btw. :)

Rebased.

**Comment by justinmk** [comment]
> On Windows there is no fsync() function but we can emulate it with _commit().
> The Windows #defines can be added later.

Any idea how much work this would be? Would prefer to address it now and remove HAVE_FYSNC.

**Comment by equalsraf** [comment]
Doesn't seem hard, defined in 7c61e19 , but untested.

Some more [pointers](https://github.com/equalsraf/neovim/commit/dbfaf9d7285bf6cc4d1d6b1910bd766764e27c3e#commitcomment-15182692) from @jasonwilliams200OK

**Comment by sethjackson** [comment]
If we wan't to get rid of `HAVE_FSYNC` now we can just use `uv_fs_fsync` and add a `os_fsync` wrapper that takes in the fd. See our `os_fchown` for an example of taking in a fd and passing it as a `uv_file`.

The Windows implementation of `uv_fs_sync` is [here](https://github.com/libuv/libuv/blob/e76b8838e51f6e8f1944b6c6d50c3948ed764a0b/src/win/fs.c#L1247).

Will update PR soonish.

**Comment by jasonwilliams200OK** [comment]
:+1: libuv's implementation looks more robust.

**Comment by sethjackson** [comment]
Travis is failing because of clint which is because of a previous change.

One question here though does that:

```
#if defined(UNIX)
```

guard need to stay?

According to the docs (`runtime/docs/options.txt`) I think not but I'm not sure...

**Comment by justinmk** [comment]
Updated the clint exceptions so this shouldn't fail for unrelated changes now.

> guard need to stay?

Technically no, but the remaining `#if defined(UNIX)` checks are reminders that a given block of code (`buf_write` especially...) needs to be explicitly compared to the upstream Vim win32 implementation and made to use our `os/` functions (and occasional `#ifdef` guards if unavoidable) in a way that makes the function work on Windows.

**Comment by sethjackson** [comment]
> Updated the clint exceptions list

Cool thanks.

> and restarted the clint build for this PR

BTW is there any way that I can do that without a `git push -f`?

As for the `#if defined(UNIX)` in this PR I guess I'll have to track that down in the Vim source then.

The other `#ifdef` maze I removed should be ok I think since we can use `fsync` everywhere now.

**Comment by ZyX-I** [comment]
@sethjackson I can go to travis, go to the clint build and press the relevant button there. But I guess this requires being at least a collaborator: e.g. in my PR to luassert I cannot do this.

File: src/nvim/shada.c [hunk 1]
```diff
@@ -761,7 +761,7 @@ static void close_sd_writer(ShaDaWriteDef *const sd_writer)
   FUNC_ATTR_NONNULL_ALL
 {
   const int fd = (int)(intptr_t) sd_writer->cookie;
-  if (fsync(fd) < 0) {
+  if (os_fsync(fd) < 0) {
     emsg2(_(SERR "System error while synchronizing ShaDa file: %s"),
           strerror(errno));
```
Comment (ZyX-I) [inline src/nvim/shada.c]: This needs to be changed: libuv errors require `uv_strerror`, `os_strerror` when wrapped by `os_*` functions.

File: src/nvim/os/fs.c [hunk 2]
```diff
@@ -202,6 +202,19 @@ int os_open(const char* path, int flags, int mode)
   return r;
 }
 
+/// Flushes file modifications to disk.
+///
+/// @param file_discriptor the file descriptor of the file to flush to disk.
```
Comment (ZyX-I) [inline src/nvim/os/fs.c]: Typo in variable name.

File: src/nvim/os/fs.c [hunk 3]
```diff
@@ -202,6 +202,19 @@ int os_open(const char* path, int flags, int mode)
...
+///
+/// @return `0` on success a libuv error code on failure.
```
Comment (ZyX-I) [inline src/nvim/os/fs.c]: Missing preposition.

**Comment by justinmk** [comment]
> As for the `#if defined(UNIX)` in this PR I guess I'll have to track that down in the Vim source then.

Out of scope for this PR IMO. But it should be just a matter of comparing the Vim Windows `buf_write` vs its Unix `buf_write`. They may be similar enough to diff like equalsraf did [here](https://github.com/neovim/neovim/pull/3304#issuecomment-144230509))

File: runtime/doc/options.txt [hunk 4]
```diff
@@ -2821,9 +2821,7 @@ A jump table for the options with a short description can be found at |Q_op|.
 	written even on filesystems which do metadata-only journaling.  This
 	will force the harddrive to spin up on Linux systems running in laptop
 	mode, so it may be undesirable in some situations.  Be warned that
-	turning this off increases the chances of data loss after a crash.  On
-	systems without an fsync() implementation, this variable is always
-	off.
+	turning this off increases the chances of data loss after a crash.
 	Also see 'swapsync' for controlling fsync() on swap files.
```
Comment (justinmk) [inline runtime/doc/options.txt]: Oh, we should remove this option.

Comment (justinmk) [inline runtime/doc/options.txt]: We should probably take this opportunity to fix `:help 'swapsync'`. E.g. this line is invalid:

```
	For a unix system
	setting it to "sync" will use the sync() call instead of the default
	fsync(), which may work better on some systems.
	The 'fsync' option is used for the actual file.
```

We might want to remove the `'swapsync'` option completely, it seems like complete overkill.

File: src/nvim/os/fs.c [hunk 5]
```diff
@@ -202,6 +202,19 @@ int os_open(const char* path, int flags, int mode)
   return r;
 }
 
+/// Flushes file modifications to disk.
+///
+/// @param file_descriptor the file descriptor of the file to flush to disk.
+///
+/// @return `0` on success, a libuv error code on failure.
+int os_fsync(int file_descriptor)
```
Comment (justinmk) [inline src/nvim/os/fs.c]: `fd` is rather idiomatic.

File: runtime/doc/options.txt (hunk 4 above)
Comment (sethjackson) [inline runtime/doc/options.txt]: :+1: to removing `swapsync`. Seems completely unnecessary.

Do that in this PR too then?

Comment (justinmk) [inline runtime/doc/options.txt]: No let's leave that for another PR.

Comment (sethjackson) [inline runtime/doc/options.txt]: Ok.

File: src/nvim/os/fs.c (hunk 5 above)
Comment (sethjackson) [inline src/nvim/os/fs.c]: Yea but we use `file_descriptor` in other places in this file so I copied that. I can use `fd` if you want and maybe update the other one too?
